File: login-system-frontend/src/services/authService.js
import apiClient from './apiClient';

/**
 * All calls to /api/v1/auth/*.
 * Returns the response data directly; callers handle errors.
 */
const authService = {
  /**
   * POST /auth/register
   * Accepts the full registration form object (name, email, password, and optional
   * phone/address fields) and returns AuthResponse { accessToken, refreshToken, user }.
   */
  register: async (formData) => {
    const { data } = await apiClient.post('/auth/register', formData);
    return data;
  },

  /**
   * POST /auth/login
   * Returns AuthResponse.
   * If requiresPasswordSet=true, tokens will be absent — caller redirects to /set-password.
   */
  login: async (email, password) => {
    const { data } = await apiClient.post('/auth/login', { email, password });
    return data;
  },

  /**
   * POST /auth/set-password  (requires Bearer token)
   * Returns AuthResponse with full tokens after account linking completes.
   */
  setPassword: async (password, confirmPassword, accessToken) => {
    const { data } = await apiClient.post(
      '/auth/set-password',
      { password, confirmPassword },
      { headers: { Authorization: `Bearer ${accessToken}` } }
    );
    return data;
  },

  /**
   * POST /auth/refresh
   * Returns a new AuthResponse with rotated tokens.
   */
  refresh: async (refreshToken) => {
    const { data } = await apiClient.post('/auth/refresh', { refreshToken });
    return data;
  },

  /**
   * POST /auth/oauth2/token
   * Exchanges the one-time OAuth code (from redirect URL) for AuthResponse.
   */
  exchangeOAuthCode: async (code) => {
    const { data } = await apiClient.post('/auth/oauth2/token', { code });
    return data;
  },
};

export default authService;

File: login-system-backend/src/main/java/com/akash/loginsystem/service/AuthService.java
package com.akash.loginsystem.service;

import com.akash.loginsystem.dto.request.LoginRequest;
import com.akash.loginsystem.dto.request.RefreshRequest;
import com.akash.loginsystem.dto.request.RegisterRequest;
import com.akash.loginsystem.dto.request.SetPasswordRequest;
import com.akash.loginsystem.dto.response.AuthResponse;
import com.akash.loginsystem.entity.User;

import java.util.UUID;

public interface AuthService {

    /** Register a new LOCAL user. Throws UserAlreadyExistsException if email taken. */
    AuthResponse register(RegisterRequest request);

    /**
     * Authenticate by email + password.
     * If user exists but passwordSet=false (OAuth user), returns AuthResponse
     * with requiresPasswordSet=true and NO tokens — client must re-authenticate via Google OAuth.
     */
    AuthResponse login(LoginRequest request);

    /**
     * Set password for the first time (account-linking flow).
     * Flips passwordSet=true. Requires authenticated user (Bearer token from OAuth redirect).
     */
    AuthResponse setPassword(UUID userId, SetPasswordRequest request);

    /**
     * Issues a full token pair for a user who authenticated via OAuth2.
     * Called by OAuth2SuccessHandler after Google login succeeds.
     */
    AuthResponse completeOAuthLogin(User user);

    /**
     * Validates the refresh token, rotates it, and issues a new access + refresh pair.
     * Throws InvalidCredentialsException if the token is unknown or expired.
     */
    AuthResponse refresh(RefreshRequest request);
}
